backend/backup_media.js

```diff
@@ -0,0 +1,151 @@
+  database: config.database
+});
+
+const entries = [];
+
+connection.query(
+  'SELECT html, "comment" as type, comment_id as "item_id" FROM `comments` WHERE `html` like "%<img%" UNION ' +
+  'SELECT html, "comment" as type, comment_id as "item_id" FROM `comments` WHERE `html` like "%<video%" UNION ' +
+  'SELECT html, "post" as type, post_id as "item_id" FROM `posts` WHERE `html` like "%<img%" UNION ' +
+  'SELECT html, "post" as type, post_id as "item_id" FROM `posts` WHERE `html` like "%<source%"',
+  async function (err, results, fields) {
+    if (err) {
+      console.trace(err);
+      process.exit(1);
+    }
+    console.log('Total items with media content: ', results.length);
+    for (let i = 0; i < results.length; i++) {
+      entries.push(results[i]);
+      if (i % 1000 === 0) {
+        console.log(`Fetched ${entries.length} entries`);
+      }
+    }
+    await startDownload();
+    process.exit(0);
+});
+
+const startDownload = async function() {
+  console.log('Starting processing');
+  for (let i = 0; i < entries.length; i++) {
+    let entry = entries[i];
+    let itemType = entry.type;
+    let itemId = entry.item_id;
+    let entryHtml = entry.html;
+
+    const mediaItems = entryHtml.match(/<(img|source).+src=['"]([^'"]+)['"]/g);
+    for (let ii = 0; ii < mediaItems.length; ii++) {
+      const check = mediaItems[ii].match(/<(img|source).+src=['"]([^'"]+)['"]/);
+      let mediaType = check[1] === MEDIA_TYPE_IMG ? MEDIA_TYPE_IMG : MEDIA_TYPE_VIDEO;
+      let mediaSrc = check[2];
+      let mediaExtensionMatch = mediaSrc.match(/\.[a-z]{2,}(?:\?.+)?$/);
+      let mediaExtension;
+      if (!mediaExtensionMatch) {
+        if (mediaType === MEDIA_TYPE_IMG) {
+          mediaExtension = '.jpg';
+        } else {
+          mediaExtension = '.mp4';
+        }
+      } else {
+        mediaExtension = mediaExtensionMatch[0].replace(/\?.+?$/, '');
+      }
+
+      let outputFilename = crypto.createHash('sha256').update(mediaSrc).digest('hex') + mediaExtension;
+      let outputPath = config.dir + '/' + outputFilename;
+      let dimensions = {width: 0, height: 0};
+
+      const saveEntry = function() {
+        if (mediaType === MEDIA_TYPE_IMG) {
+          try {
+            dimensions = sizeOf(outputPath);
+          } catch (e) {
+            dimensions = {width: 0, height: 0};
+          }
+        }
+        let line = [
+          itemType, itemId, mediaType, mediaSrc, outputFilename, dimensions.width, dimensions.height
+        ];
+        fs.appendFileSync(config.dbFile, line.join(',') + "\n");
+      }
+
+      if (fs.existsSync(outputPath)) {
+        console.log(outputPath, 'is already there');
+        saveEntry();
+        continue;
+      }
+      download(mediaSrc, outputPath, async function() {
+        saveEntry();
+        console.log('Downloaded:', itemType, itemId, mediaType, mediaSrc, config.dir + '/' + outputFilename, dimensions.width, 'x', dimensions.height);
+      });
+      await sleep(500);
```

Comment:
Do I understand correctly, that the main loop just spawns download jobs every 500 ms without waiting for their completion? Is that intentional?

Reply:
yes, but this is not on purpose. I'll make it synchronous